include/runtime/proof_trace_writer.h

```diff
@@ -0,0 +1,142 @@
+#ifndef RUNTIME_PROOF_TRACE_WRITER_H
+#define RUNTIME_PROOF_TRACE_WRITER_H
+
+#include <kllvm/binary/ProofTraceParser.h>
+#include <runtime/header.h>
+
+#include <cstdio>
+
+class proof_trace_writer {
+public:
+  virtual ~proof_trace_writer() = default;
+
+  virtual void proof_trace_header(uint32_t version) = 0;
+  virtual void hook_event_pre(
+      char const *name, char const *pattern, char const *location_stack)
+      = 0;
+  virtual void
+  hook_event_post(void *hook_result, uint64_t block_header, bool indirect)
+      = 0;
+  virtual void argument(void *arg, uint64_t block_header, bool indirect) = 0;
+  virtual void rewrite_event_pre(uint64_t ordinal, uint64_t arity) = 0;
+  virtual void
+  variable(char const *name, void *var, uint64_t block_header, bool indirect)
+      = 0;
+  virtual void
+  rewrite_event_post(void *config, uint64_t block_header, bool indirect)
+      = 0;
+  virtual void function_event_pre(char const *name, char const *location_stack)
+      = 0;
+  virtual void function_event_post() = 0;
+  virtual void side_condition_event_pre(uint64_t ordinal, uint64_t arity) = 0;
+  virtual void
+  side_condition_event_post(uint64_t ordinal, bool side_cond_result)
+      = 0;
+  virtual void configuration(block *config) = 0;
+  virtual void end_of_trace() = 0;
+};
+
+class proof_trace_file_writer : public proof_trace_writer {
+private:
+  FILE *file_;
+
+  void write(void const *ptr, size_t len) { fwrite(ptr, len, 1, file_); }
+
+  void write_string(char const *str, size_t len) { fwrite(str, 1, len, file_); }
+
+  // Note: This method will not write a 0 at the end of string.
+  // The passed string should be 0 terminated.
+  void write_string(char const *str) { fputs(str, file_); }
+
+  // Note: this method will write a 0 at the end of the string.
+  // The passed string should be 0 terminated.
+  void write_null_terminated_string(char const *str) {
+    write_string(str);
+    char n = 0;
+    write(&n, 1);
+  }
```

Comment:
It would be clearer to rename write() to something more informative; say write_bytes(). Otherwise it looks to much like the system call write().

Reply:
I think I will go with @stevenmeker suggestion here. Calling `write_string` which is documented to not add a null at the end with the null string does not seem very clear to me. I will use `\0` though instead of `0` as @Robertorosmaninho suggested.

Reply:
Sure! I'm happy with it too!